### ISSUE TITLE
Questions about QSM

### ISSUE THREAD

**Issue by wadx2019** [Alescontrela/score_matching_rl#1]
Hello, authors of QSM. I am a master student working on RL and have an issue on the proposed method.

Why we cannot directly utilize the gradient of Q function but train an extra score model to do the denoising procedure? I think the approximation error of score matching can be avoided in this way.

**Comment by xiaohangt** [comment]
I have exactly the same question, what is the author's opinion on this please? Thank you very much! @Alescontrela @michael-psenka 

Are you trying to avoid extensive autodiff during denoising? Or you have any other considerations?

**Comment by michael-psenka** [comment]
Thanks for the question! Note the paper's theory applies to whether or not the policy score is set to exactly the "Q-score" (\nabla_a Q) or trained to iteratively become closer. It shouldn't be thought of as an approximation error or not (this would be true if we had immediate access to the optimal Q^*, but this needs to be learned too), but rather whether the training process is "decoupled" (like QSM, think akin to student-teacher networks), or "coupled" (like what you all propose). The theory shows that this matching can be thought of as a functional "gradient update", and how close we match the denoising model to the Q-score each iteration should be thought of as akin to a learning rate for gradient descent; perfectly valid to try making it higher, but not necessarily correct/optimal in every scenario.

What you describe though -- only training a Q-network, and sampling essentially from adjusted Langevin dynamics -- is a valid critic-only approach, and one we actually did some experiments on midway through, and it definitely works decently well. We decided for this paper to focus on the full actor-critic setting with separate networks namely to stay closer to the more standard actor-critic framework, but also to keep the time-dependency throughout the denoising process and to demonstrate that the theory for QSM applies to a more general array of diffusion model policy setups. We didn't test the exploration capabilities of this critic-only variant, but we may lose the natural convergence to multimodal policies in the critic-only variant? Would be an interesting experiment to explore.

**td;lr**, what you propose is a valid alternative, but it's not necessarily a strict improvement because it's not necessarily better for training dynamics to immediately jump the policy to the current \nabla_a Q _\theta every step. The QSM theory allows for many possible ways of training diffusion model policies from rewards, our implementation is just one of them. The proposed alternative an alternative I'm quite interested in though, as the theory very much hints towards it, and am excited to see the research to come.

**Comment by michael-psenka** [comment]
I'll close the issue since it doesn't pertain to code issues, but if either of you have more questions, please feel free to reach out via email (you can find this on the paper's arxiv).
